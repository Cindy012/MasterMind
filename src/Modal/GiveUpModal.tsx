interface GiveUpModalProps {
    setShowModal: React.Dispatch<React.SetStateAction<boolean>>;
    hideCloseButton?: boolean;
    show: boolean;
    showAnswer: () => void;
    title: string;
}

const GiveUpModal:React.FC<GiveUpModalProps> = ({ setShowModal, show, hideCloseButton, showAnswer, title }) => {
    const stopGame = () => {
        showAnswer();
        setShowModal(false);
    };

    return (
        <div className={`modal ${show ? 'active' : ''}`}>
            <div className='modal__content'>
                { !hideCloseButton && <span onClick={() => setShowModal(false)} className="modal__close">&times;</span> }
                <h2>{ title }</h2> 
                <p>Are you sure to give up?</p>
                <div className="modal__footer">
                    <button className="modal__button" onClick={ () => stopGame() }>Yes</button>
                    <button className="modal__button" onClick={ () => setShowModal(false) }>No</button>
                    
                </div>
            </div>
        </div>
    );
};

export default GiveUpModal;
