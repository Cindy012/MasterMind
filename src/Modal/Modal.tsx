import confetti from 'canvas-confetti';
import { useEffect, useState } from 'react';

interface ModalProps {
    setShowModal: React.Dispatch<React.SetStateAction<boolean>>;
    hideCloseButton?: boolean;
    show: boolean;
    resetGame?: () => void
    gameStatus: number; // 0: gameInfo, 1: not filled/ongoing, 2: winner, 3: loser
    gameInfo?: JSX.Element;
}

const Modal:React.FC<ModalProps> = ({ setShowModal, show, hideCloseButton, resetGame, gameStatus, gameInfo }) => {
    const [playConfetti, setPlayConfetti] = useState(false);
    const playAgain = () => {
        if (resetGame) {
            setPlayConfetti(false);
            resetGame();
            setShowModal(false);
        }
    };

    const title: string[] = [
        'Mastermind game info',
        'Not Yet!',
        'Winner!',
        'Loser!'
    ];

    const text: string[] = [
        '',
        'You did not give the code colors.',
        'You win, play again!',
        'You lose, next time better!'
    ];

    if (gameStatus === 2 && resetGame && !playConfetti) {
        setPlayConfetti(true);
    }

    useEffect(() => {
        if (playConfetti) {
            confetti({
                particleCount: 160,
                spread: 100,
                colors: ['#e43b3b', '#e48a3b', '#e4de3b', '#3be45d', '#3be4de', '#3b60e4', '#8d3be4', '#e43b9b']
            });
        }
    },[playConfetti]);
       
    return (
        <div className={`modal ${show ? 'active' : ''}`}>
            <div className={ gameInfo ? 'modal__content' : 'modal__content modal__content__small' }>
                { !hideCloseButton && <span onClick={() => setShowModal(false)} className="modal__close">&times;</span> }
                <h2>{ title[gameStatus] }</h2> 
                { gameInfo && gameStatus === 0 ? gameInfo : <p>{ text[gameStatus] }</p> }
                <div className={ gameStatus === 0 || gameStatus === 1 ? 'modal__footer__secondary' : 'modal__footer'}>
                    { gameStatus !== 1 && gameStatus !== 0 ? (
                        <button className="modal__button" onClick={ () => playAgain() }>Play again!</button>
                    ) : null }
                    <button className="modal__button" onClick={ () => setShowModal(false) }>Close</button>
                </div>
            </div>
        </div>
    );
};

export default Modal;
