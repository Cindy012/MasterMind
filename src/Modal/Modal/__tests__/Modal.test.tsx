/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react';
import '@testing-library/jest-dom';
import Modal from '../Modal';
import { render } from 'react-dom';

describe('Modal', () => {
    it('renders the (gameInfo) Modal successfully and is not active', () => {
        const setStateMocked = jest.fn();
        const useStateMock: any = (useState: any) => [useState, setStateMocked];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        const [showModalMock, setShowModalMock] = useStateMock(false);

        const container = document.createElement('div');
        render(<Modal setShowModal={ setShowModalMock } show={ showModalMock } gameStatus={ 0 } />, container);
        expect(container).toMatchSnapshot();
    });

    it('renders the (gameInfo) Modal successfully and is active', () => {
        const setStateMocked = jest.fn();
        const useStateMock: any = (useState: any) => [useState, setStateMocked];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        const [showModalMock, setShowModalMock] = useStateMock(true);

        const container = document.createElement('div');
        render(<Modal setShowModal={ setShowModalMock } show={ showModalMock } gameStatus={ 0 } />, container);
        expect(container).toMatchSnapshot();
    });
});