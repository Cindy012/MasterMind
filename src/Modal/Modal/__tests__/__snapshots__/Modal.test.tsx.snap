// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders the gameInfo Modal successfully and is not active 1`] = `
<div>
  <div
    class="modal "
    data-testid="modal"
  >
    <div
      class="modal__content null"
    >
      <span
        class="modal__close"
      >
        Ã—
      </span>
      <h2>
        Mastermind game info
      </h2>
      <article>
        <p>
          MasterMind is a game whereby the user guesses code on four pegs. The player has eight chances to guess the color code.
        </p>
        <p>
          To choose a color, select from the side one of the colors pegs. After selecting the color, the player can fill the row of the board.
        </p>
        <p>
          When the player selects a code pattern, the code will be reviewed and provides feedback through clues. The black peg means that the chosen peg color and position are correct. The red peg means that the peg color retains in the code although the position is not.
        </p>
      </article>
      <div
        class="modal__footer__secondary"
      >
        <button
          class="modal__button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
